src/main/groovy/com/lesfurets/jenkins/unit/BasePipelineTest.groovy

```diff
@@ -25,13 +25,30 @@ abstract class BasePipelineTest {
 
     def stringInterceptor = { m -> m.variable }
 
+    def usernamePasswordInterceptor = { m -> [m.usernameVariable, m.passwordVariable] }
+
     def withCredentialsInterceptor = { list, closure ->
-        list.forEach {
-            binding.setVariable(it, "$it")
+        list.forEach { creds ->
+            if (creds instanceof String) {
+                binding.setVariable(creds, creds)                
```

Comment:
I think it is a good idea to save previous value if any and then to restore it, 
im not sure if Jenkins does support nested calls but in your case
```
withCredentails(usernamePassword(['user', 'passwd'])) {
  sh 'somescript $user $passwd' // ok
  withCredentails(usernamePassword(['user', 'passwd'])) {
     sh 'another-script $user $passwd' // ok
  }
  sh 'somescript $user $passwd' // values would be null
}
```

Reply:
could you please add a tests to verify credentails restore procedure

Reply:
I think this is probably an odd case but the original code did not do that. Since I am already in this area I might as well fix that. The same could be said for the withEnv code which I also changed to restore the environment by removing the added values.

Reply:
Done